### PR TITLE
Stitch1DMany new property 'ScaleFactorFromPeriod'

### LINKED ISSUE #17819: `Stitch1DMany` should be able to apply scale factor from a particular period
Requested by scientists for release 3.9.

From Max:

> `Stitch1DMany` applies different scale factor to each period. We'd like an option to be able to determine the scale factor from a particular period and then apply the same on to all others. We still want to be able to calculate different ones for each period.

### PR DESCRIPTION
**Pull request by PranavBahuguna**
When stitching workspaces from multiple groups together, the algorithm currently allows a global scale factor to be specified using properties `UseManualScaleFactor` and `ManualScaleFactor`. This new property `ScaleFactorFromPeriod` allows a set of global scale factors to be determined from workspaces in groups from a specific period and applies them to workspaces in all other periods.

This property is only used when `UseManualScaleFactor` is true and `ManualScaleFactor` is set to its default value. Furthermore, the documentation regarding the `Stitch1DMany` algorithm has been updated and improved.

**To test:**

- Ensure all unit tests pass.
- Ensure that the `ScaleFactorFromPeriod` property accomplishes what it was meant to do.
- Ensure that the documentation is up to date and correct.

Fixes #17819.

*Release notes have been updated.*

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/Stitch1DMany.cpp [hunk 1]
```diff
@@ -113,151 +176,248 @@ std::map<std::string, std::string> Stitch1DMany::validateInputs() {
...
+  alg->execute();
+
+  if (!isChild()) {
+    // Copy each input workspace's history into our output workspace's history
+    for (auto &inputWS : m_inputWSMatrix[wsIndex])
+      lhsWS->history().addHistory(inputWS->getHistory());
+  }
+  // We're a child algorithm, but we're recording history anyway
+  else if (isRecordingHistoryForChild() && m_parentHistory) {
+    m_parentHistory->addChildHistory(m_history);
```
Comment (raquelalvarezbanos) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: If I run `Stitch1DMany` on matrix workspaces (i.e. no groups):

```
x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
y = [1, 1, 1, 1, 1, 1, 1, 1, 1]
ws_1_1 = CreateWorkspace(x, y)

x = [9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
y = [2, 2, 2, 2, 2, 2, 2, 2, 2]
ws_2_1 = CreateWorkspace(x, y)

x = [17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
y = [3, 3, 3, 3, 3, 3, 3, 3, 3]
ws_3_1 = CreateWorkspace(x, y)

stitched, scale_factors = Stitch1DMany(InputWorkspaces='ws_1_1, ws_2_1, ws_3_1', Params='1')
```

On master the history contains also `CreateWorkspace v.1`, but on this branch only `Stitch1DMany v.1` appears.

File: Framework/Algorithms/src/Stitch1DMany.cpp [hunk 2]
```diff
@@ -113,151 +176,248 @@ std::map<std::string, std::string> Stitch1DMany::validateInputs() {
...
-
   if (!m_scaleRHSWorkspace) {
     // Flip these around for processing
-    std::reverse(m_inputWorkspaces.begin(), m_inputWorkspaces.end());
     std::reverse(m_startOverlaps.begin(), m_startOverlaps.end());
     std::reverse(m_endOverlaps.begin(), m_endOverlaps.end());
+    for (auto &period : m_inputWSMatrix) {
+      std::reverse(period.begin(), period.end());
+    }
   }
```
Comment (raquelalvarezbanos) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: I am not sure this `if (m_scaleRHSWorkspace)` block is needed. Don't change this for the moment but keep an eye on https://github.com/mantidproject/mantid/pull/18162 to see if the tester agrees it is not needed. In that case, we should remove it here too.

**Review by raquelalvarezbanos — COMMENTED** [review]
I have some suggestions (see below). In addition, I found the following:

When running this script:

```
import numpy as np

def straight_line(x, a, b):
  return a+b*x

#create two histograms
x1 = np.arange(-1, 1, 0.02)
x2 = np.arange(0.4, 1.6, 0.02)
x3 = np.arange(1.3, 3, 0.02)
ws_1_1 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_1 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 1, 1)+1)
ws_3_1 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 1, 1)+1)
ws_1_2 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_2 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 1.5, 1)+1)
ws_3_2 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 2, 1)+1)
ws_1_3 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_3 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 2, 1)+1)
ws_3_3 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 3, 1)+1)
group1 = GroupWorkspaces('ws_1_1, ws_1_2, ws_1_3')
group2 = GroupWorkspaces('ws_2_1, ws_2_2, ws_2_3')
group3 = GroupWorkspaces('ws_3_1, ws_3_2, ws_3_3')

stitched_from_period, factors = Stitch1DMany('group1, group2, group3', Params=[0.02], UseManualScaleFactor=True, ScaleFactorFromPeriod=1)
print factors

# Lines below should produce the same result as the line above
stitched_1, scale_factor_1 = Stitch1D('ws_1_2', 'ws_2_2')
stitched_1, scale_factor_2 = Stitch1D(stitched_1, 'ws_3_2')
stitched_0, sf1 = Stitch1D('ws_1_1', 'ws_2_1', UseManualScaleFactor=True, ManualScaleFactor=scale_factor_1)
stitched_0, sf2 = Stitch1D(stitched_0, 'ws_3_1', UseManualScaleFactor=True, ManualScaleFactor=scale_factor_2)
stitched_2, sf1 = Stitch1D('ws_1_3', 'ws_2_3', UseManualScaleFactor=True, ManualScaleFactor=scale_factor_1)
stitched_2, sf2 = Stitch1D(stitched_2, 'ws_3_3', UseManualScaleFactor=True, ManualScaleFactor=scale_factor_2)
stitched = GroupWorkspaces('stitched_0, stitched_1, stitched_2')
```

if I compare workspaces `stitched` and `stitched_from_period` the almost look the same, which is good, but there's a bin that differs. I haven't had time to look into it, could you have a look? it could be that I am doing something wrong in the script.

**Comment by raquelalvarezbanos** [comment]
Regarding the problem I found above: I think it has nothing to do with your changes. If I do:

```
import numpy as np

def straight_line(x, a, b):
  return a+b*x

#create two histograms
x1 = np.arange(-1, 1, 0.02)
x2 = np.arange(0.4, 1.6, 0.02)
x3 = np.arange(1.3, 3, 0.02)
ws_1_1 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_1 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 1, 1)+1)
ws_3_1 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 1, 1)+1)
ws_1_2 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_2 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 1.5, 1)+1)
ws_3_2 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 2, 1)+1)
ws_1_3 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=straight_line(x1[:-1], 1, 1)+1)
ws_2_3 = CreateWorkspace(UnitX="1/q", DataX=x2, DataY=straight_line(x2[:-1], 2, 1)+1)
ws_3_3 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=straight_line(x3[:-1], 3, 1)+1)

stitched_1 = Stitch1DMany('ws_1_2, ws_2_2, ws_3_2', Params=[0.02])
stitched_2, scale_factor_1 = Stitch1D('ws_1_2', 'ws_2_2')
stitched_2, scale_factor_2 = Stitch1D(stitched_2, 'ws_3_2')
```

on master, I get the same problem (`stitched_1` and `stitched_2` are the same except for one bin that differs). The problem seems to be that when running `Stitch1D` I am not passing any value for `Params`, if I modify the script above like this:

```
```

File: Framework/Algorithms/src/Stitch1DMany.cpp (hunk 1 above)
Comment (raquelalvarezbanos) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: Regarding the first script: I found one way to solve it (not sure it's the best thing to do). If we do `alg->setChild(false)` in `doStitch1DMany()` the output workspace has a history attached. We'll have to check if this has any effect on the other cases (i.e. the case with no groups and the case when workspaces are stitched from a specific period).

Comment (PranavBahuguna) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: As far as I can tell, this does not appear to cause any side effects and causes the history to display correctly. I will include this.

File: Framework/Algorithms/src/Stitch1DMany.cpp (hunk 2 above)
Comment (raquelalvarezbanos) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: Seems that there's no need to do the above `if` (see my other PR). Can you remove it and check the effect this has? I think with this `if` the algorithm fails if `ScaleRHSWorkspace = false`, without it, it should work and results should make sense. Could you check this?

Comment (raquelalvarezbanos) [inline Framework/Algorithms/src/Stitch1DMany.cpp]: OK, thanks. When you remove this, could you try a couple of sample scripts and check that the result is as you'd expect?

**Comment by raquelalvarezbanos** [comment]
I showed the new property to Christy and he is happy with how it works (although he will have to test it on real data). He only suggested if the period number could range from `1` to `N`, instead of the current `0` to `N-1`. Can we change this? I think this should be the last change and this PR can be merged.

**Comment by raquelalvarezbanos** [comment]
New functionality looks good. The only problem is that the code is getting a bit complicated, I need to check with scientists if we can replace `ManualScaleFactor` with a vector (that would make things a lot easier). I've created a maintenance issue (https://github.com/mantidproject/mantid/issues/18378) for this. Other than that, everything works as expected.

:shipit:

**Comment by raquelalvarezbanos** [comment]
@PranavBahuguna there's a conflict that needs to be resolved.

**Pull request merged**